src/main/scala/edu/ie3/osmogrid/lv/coordinator/LvCoordinator.scala

```diff
@@ -0,0 +1,227 @@
+        stopBehavior
+    }
+    .receiveSignal { case (ctx, PostStop) =>
+      postStopCleanUp(ctx.log)
+    }
+
+  /** Await incoming input data and register it
+    *
+    * @param awaitingData
+    *   State data for the awaiting
+    * @return
+    *   Equivalent next state
+    */
+  private def awaitInputData(
+      awaitingData: AwaitingData
+  ): Behavior[Request] = Behaviors
+    .receive[Request] {
+      case (ctx, MessageAdapters.WrappedInputDataResponse(response)) =>
+        /* Register what has been responded */
+        awaitingData.registerResponse(response, ctx.log) match {
+          case Success(updatedStateData) =>
+            handleUpdatedAwaitingData(updatedStateData, ctx)
+          case Failure(exception) =>
+            ctx.log.error(
+              "Request of needed input data failed. Stop low voltage grid generation.",
+              exception
+            )
+            stopBehavior
+        }
+      case (ctx, Terminate) =>
+        terminate(ctx.log)
+      case (ctx, unsupported) =>
+        ctx.log.warn(
+          s"Received unsupported message '$unsupported' in data awaiting state. Keep on going."
+        )
+        Behaviors.same
+    }
+    .receiveSignal { case (ctx, PostStop) =>
+      postStopCleanUp(ctx.log)
+    }
+
+  /** Handle updated [[AwaitingData]]. If everything, that is requested, is at
+    * place, spawn child actors and change to Behavior to await results. If
+    * still something is missing, wait for it.
+    *
+    * @param awaitingData
+    *   Updated state data
+    * @param ctx
+    *   Actor context to use
+    * @return
+    *   Next state
+    */
+  private def handleUpdatedAwaitingData(
+      awaitingData: AwaitingData,
+      ctx: ActorContext[Request]
+  ): Behavior[Request] = {
+    /* Check, if everything is in place */
+    if (
+      awaitingData.osmData.isDefined && awaitingData.assetInformation.isDefined
+    ) {
+      /* Process the data */
+      ctx.log.debug("All awaited data is present. Start processing.")
+
+      /* Spawn an coordinator for the region */
+      ctx.self ! StartGeneration(
+        awaitingData.cfg,
+        ctx.spawnAnonymous(
+          LvRegionCoordinator()
+        )
+      )
+
+      /* Wait for results to come up */
+      awaitResults(awaitingData.guardian, awaitingData.msgAdapters)
+    } else
+      awaitInputData(awaitingData) // Wait for missing data
+  }
+
+  /** State to receive results from subordinate actors
+    *
+    * @param guardian
+    *   Reference to the guardian actor
+    * @param msgAdapters
+    *   Collection of all apparent message adapters
+    * @return
+    *   The next state
+    */
+  private def awaitResults(
+      guardian: ActorRef[Response],
+      msgAdapters: MessageAdapters
+  ): Behavior[Request] = Behaviors
+    .receive[Request] {
+      case (ctx, StartGeneration(cfg, regionCoordinator)) =>
+        /* Forward the generation request */
+        regionCoordinator ! LvRegionCoordinator.Partition(
```

Comment:
Isn't the goal that an actor only sends messages of `their own api` and the receiving actor wrap them with these message adapters?

Reply:
Not to my knowledge. The messages that are understood by an agent somewhat define the "API" of this agent, so which functions the agent might serve -- somewhat like a method definition. If you would put this to the sending actor, you would distribute the definition of possible action throughout the whole code. Moreover, if two distinct actors want to do the same request (content-wise), you'd have to implement that twice in the receiving actor.